libs/knowledge-store/ragstack_knowledge_store/base.py

```diff
@@ -0,0 +1,345 @@
+from langchain_core.load import Serializable
+from langchain_core.runnables import run_in_executor
+from langchain_core.vectorstores import VectorStore, VectorStoreRetriever
+from pydantic import Field
+
+
+class Node(Serializable):
+    """Node in the KnowledgeStore graph"""
+
+    id: Optional[str]
+    """Unique ID for the node. Shall be generated by the KnowledgeStore if not set"""
+    metadata: dict = Field(default_factory=dict)
+    """Metadata for the node. May contain information used to link this node 
+    with other nodes."""
+
+
+class TextNode(Node):
+    text: str
+    """Text contained by the node"""
+
+
+def _texts_to_nodes(
+    texts: Iterator[str],
+    metadatas: Iterator[dict],
+    ids: Iterator[str],
+) -> Iterator[Node]:
+    for text, metadata, node_id in zip(texts, metadatas, ids):
+        yield TextNode(
+            id=node_id,
+            metadata=metadata,
+            text=text,
+        )
+
+
+def _documents_to_nodes(
+    documents: Iterator[Document], ids: Iterator[str]
+) -> Iterator[Node]:
+    for (
+        doc,
+        node_id,
+    ) in zip(documents, ids):
+        yield TextNode(
+            id=node_id,
+            metadata=doc.metadata,
+            text=doc.page_content,
+        )
+
+
+class KnowledgeStore(VectorStore):
+    """A hybrid vector-and-graph knowledge store.
+
+    Document chunks support vector-similarity search as well as edges linking
+    chunks based on structural and semantic properties.
+    """
+
+    @abstractmethod
+    def add_nodes(
+        self,
+        nodes: Iterable[Node],
+        **kwargs: Any,
+    ) -> List[str]:
+        """Add nodes to the knowledge store
+
+        Args:
+            nodes: the nodes to add.
+        """
+
+    async def aadd_nodes(
+        self,
+        nodes: Iterable[Node],
+        **kwargs: Any,
+    ) -> List[str]:
+        """Add nodes to the knowledge store
+
+        Args:
+            nodes: the nodes to add.
+        """
+        return await run_in_executor(None, self.add_nodes, nodes, **kwargs)
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[Iterable[dict]] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        metadatas_it = iter(metadatas) if metadatas else itertools.repeat({})
+        ids_it = iter(ids) if ids else itertools.repeat(None)
+        nodes = _texts_to_nodes(iter(texts), metadatas_it, ids_it)
+        return self.add_nodes(nodes, **kwargs)
+
+    async def aadd_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[Iterable[dict]] = None,
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        metadatas_it = iter(metadatas) if metadatas else itertools.repeat({})
+        ids_it = iter(ids) if ids else itertools.repeat(None)
+        nodes = _texts_to_nodes(iter(texts), metadatas_it, ids_it)
```

Comment:
May want `_texts_to_nodes` to take the iterables and do the repeat. Specifically, there is the following issue:

1. We'd like to ensure that if presented, the metadatas, ids and texts all have the same lengths. This suggests `zip(texts, metadatas, ids, strict=True)`.
2. That will fail when presented the infinite repeat iterators.
3. So `_texts_to_nodes` will need to look at which of them are present to decide which to zip over.

Reply:
`strict` was added to Python 3.10 so I think we should refrain from using it atm.
LangChain is compatible with Python 3.8.1+

Reply:
Done (removing zip and checking iterators manually)